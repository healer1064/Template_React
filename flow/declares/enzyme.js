// @flow

declare module 'enzyme' {

  declare var shallow: Function

}
