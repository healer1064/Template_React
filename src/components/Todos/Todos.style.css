.profile {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  color: green;

  .modal {
    display: flex;
    flex-direction: column;
    max-width: 300px;

    button {
      margin-top: 1rem;
    }
  }
}
