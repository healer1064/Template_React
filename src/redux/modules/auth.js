// @flow
import { setAuthToken } from 'src/helpers/auth'
import {
  loginMock as loginApi,
} from 'src/helpers/api/auth'
import history from 'src/helpers/history'

export const LOGIN = 'my-app/auth/LOGIN'

// Flow type for this reducer's initial state
export type AuthState = {
  authenticated: boolean,
  user: Object,
  error: ?string,
  isFetching: boolean,
}

// Initial state with default values
export const initialState = {
  authenticated: false,
  user: {},
  error: '',
  isFetching: false,
}


// REDUCER
function reducer (state: AuthState = initialState, action: GlobalFSA<*>) {

  switch (action.type) {

    case `${LOGIN}_PENDING`:
      return {
        ...state,
        isFetching: true,
      }

    case `${LOGIN}_FULFILLED`:
      return {
        ...state,
        authenticated: true,
        error: '',
        isFetching: false,
      }

    case `${LOGIN}_REJECTED`:
      return {
        ...state,
        authenticated: false,
        error: action.payload,
        isFetching: false,
      }

    default:
      return state

  }

}


// ACTION CREATORS
// Use redux-promise-middleware
// https://github.com/pburtchaell/redux-promise-middleware/blob/master/docs/guides/chaining-actions.md
// Which, in turn, uses Flux Standard Action (FSA) notation
// https://github.com/acdlite/flux-standard-action

export const login = (
  username: string, password: string, redirect?: string,
) => {

  return (dispatch: any) => {

    return dispatch({
      type: LOGIN,
      payload: loginApi({ username, password })
        .then((data) => {

          setAuthToken(data.account.jwt)
          history.push({
            pathname: redirect,
          })

        }),
    })

  }

}


export default reducer
