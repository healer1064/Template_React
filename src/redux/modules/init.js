// @flow
export const LOAD = 'my-app/init/LOAD'
export const LOAD_SUCCESS = 'my-app/init/LOAD_SUCCESS'

// import { Map } from 'immutable'
// import Debug from 'debug'

// const log = Debug('my-app')


type InitAction = {
  type: ?string,
}

type InitState = {
  isLoading: boolean,
  loaded: boolean,
}


export const initialState: InitState = {
  isLoading: true,
  loaded: false,
}


function reducer (state: ?InitState = initialState, action: InitAction)
  : ?InitState {

  switch (action.type) {


    case LOAD:
      return {
        ...state,
        isLoading: true,
      }

    case LOAD_SUCCESS:
      return {
        ...state,
        isLoading: false,
        loaded: true,
      }


    default:
      return state

  }

}


// Removed from implementation until the need arises
// For now, we assume that the app arrives in loading state
export function load (): InitAction {

  return {
    type: LOAD,
  }

}

export function loadSuccess (): InitAction {

  return {
    type: LOAD_SUCCESS,
  }

}


export default reducer
