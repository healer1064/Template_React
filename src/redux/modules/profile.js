// @flow
import { Map, Record } from 'immutable'
import { Observable } from 'rxjs/Observable'
import axios from 'axios'


export const FETCH_PROFILE: string = 'my-app/profile/FETCH_PROFILE'
export const FETCH_PROFILE_SUCCESS: string = 'my-app/profile/FETCH_PROFILE_SUCCESS'
export const FETCH_PROFILE_ERROR: string = 'my-app/profile/FETCH_PROFILE_ERROR'


// Flow type for this model
// add "T" to the end if there is a naming colision
type ProfileT = Record<{
  id: number,
  firstName: string,
  lastname: string,
  email: string,
  city: string,
  dob: string,
  notifications: boolean,
  picture: string,
}>

// Flow type for this reducer's state
export type ProfileState = Record<{
  me: number,
  list: Map<[[number, ProfileT]]>,
}>

// Flow type for this reducer's action
// Use Flux Standard Action (FSA) notation
// https://github.com/acdlite/flux-standard-action
// However we alway pass the payload. If empty, pass {} to avoid runtime errors
type PayloadFSA = {}
type FetchPayloadFSA = {
  url: string,
}

// Profile model with default values
export const Profile = Record({
  id: 1,
  firstName: '',
  lastName: '',
  email: '',
  city: 'Nashville, TN',
  dob: '1970-01-01T00:00:00.000Z',
  notifications: true,
  picture: '',
})


// Initial state with default values
const InitialState = Record({
  me: 1,
  list: Map([
    [1, new Profile()], // Use tuple to ensure numbers are always used for ids
  ]),
})


export const initialState = new InitialState()
// Now we can retrieve city with initialState.getIn(['list', 1, 'city'])
// Or my city with initialState.getIn(['list', initialState.get('me'), 'city'])


function reducer (
  state?: ProfileState = initialState,
  action: GlobalFSA<PayloadFSA>
): ProfileState {

  switch (action.type) {

    default:
      return state

  }

}


// Action Creators
export const fetchProfile = (payload: PayloadFSA): GlobalFSA<PayloadFSA> => {

  return {
    type: FETCH_PROFILE,
    payload,
  }

}

const fetchProfileSuccess = (): GlobalFSA<PayloadFSA> => {

  return {
    type: FETCH_PROFILE_SUCCESS,
    payload: {},
  }

}

const fetchProfileError = (): GlobalFSA<PayloadFSA> => {

  return {
    type: FETCH_PROFILE_ERROR,
    payload: {},
    error: true,
  }

}
// Always append error: true for redux action error types


// Redux-Observable Epics
// variable$ notation indicates an event stream
export const fetchProfileEpic = (action$: Observable) => {

  action$.ofType(FETCH_PROFILE)
    .mergeMap((action: GlobalFSA<FetchPayloadFSA>) => {

      axios.get(action.payload.url)

    })
    .map(fetchProfileSuccess)
    .catch(fetchProfileError)

}


export default reducer
