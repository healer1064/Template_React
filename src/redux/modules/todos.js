// @flow
// Non-shallow reducer state example needs Immutable
// Async actions need redux-observable epics
import { List, Record } from 'immutable'
import {
  getTodosMock as getTodosApi,
} from '../../helpers/api/examples/todos'


// ACTION TYPES
// export const ADD_TODO = 'my-app/todo/ADD_TODO'
// export const COMPLETE_TODO = 'my-app/todo/COMPLETE_TODO'
export const GET_TODOS = 'my-app/todos/GET_TODOS'
export const SET_FILTER = 'my-app/todo/SET_FILTER'

// FILTERS
export const FILTER_CURRENT = 'my-app/todo/FILTER_CURRENT'
export const FILTER_DONE = 'my-app/todo/FILTER_DONE'


// Flow type for this reducer's state
// Always put the asterisk or it assumes 'empty'
export type TodosState = Record<*>

// MODEL
// Profile model with default values
export class Todo extends Record({ text: '', date: '', done: false }) {

  constructor ({
    text,
    date,
    done,
  }: { text?: string, date?: string, done?: boolean }) {

    super({
      text,
      date: date || new Date(),
      done: done || false,
    })

  }

}

// Initial state with default values
class InitialState extends Record({
  filter: FILTER_CURRENT,
  list: List(),
  isFetching: false,
  error: '',
}) {

  constructor ({
    filter,
    list,
    isFetching,
    error,
  }: {
    filter?: string,
    list?: Array<*>,
    isFetching?: boolean,
    error?: string,
  } = {}) {

    super({
      filter,
      list,
      isFetching,
      error,
    })

  }

}


const parseApiTodoList = (data) => {

  if (!data || data.length === 0) return List()

  return data.map((item) => {

    return new Todo(item)

  })

}


export const initialState = new InitialState()
// Now we can retrieve city with initialState.getIn(['list', 1, 'city'])
// Or my city with initialState.getIn(['list', initialState.get('me'), 'city'])


// REDUCER
function reducer (state: TodosState = initialState, action: GlobalFSA<*>) {

  switch (action.type) {

    case SET_FILTER:
      return state.set('filter', action.payload.filter)

    case `${GET_TODOS}_PENDING`:
      return state
        .set('isFetching', true)

    case `${GET_TODOS}_FULFILLED`:
      return state
        .set('list', parseApiTodoList(action.payload))
        .set('error', '')
        .set('isFetching', false)

    case `${GET_TODOS}_REJECTED`:
      return state
        .set('error', action.payload)
        .set('isFetching', false)

    default:
      return state

  }

}


// ACTION CREATORS
// Use redux-promise-middleware
// https://github.com/pburtchaell/redux-promise-middleware/blob/master/docs/guides/chaining-actions.md
// Which, in turn, uses Flux Standard Action (FSA) notation
// https://github.com/acdlite/flux-standard-action
export const getTodos = () => {

  return (dispatch: any) => {

    return dispatch({
      type: GET_TODOS,
      payload: getTodosApi(),
    })

  }

}


// Plain actions uses Flux Standard Action (FSA) notation
// https://github.com/acdlite/flux-standard-action
export const setFilterDone = () => {

  return {
    type: SET_FILTER,
    payload: {
      filter: FILTER_DONE,
    },
  }

}


export const setFilterCurrent = () => {

  return {
    type: SET_FILTER,
    payload: {
      filter: FILTER_CURRENT,
    },
  }

}


// EPICS
// variable$ notation indicates an event stream
// https://redux-observable.js.org/docs/basics/Epics.html


export default reducer
