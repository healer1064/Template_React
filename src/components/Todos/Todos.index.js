// @flow
import type { List } from 'immutable'
import { connect } from 'react-redux'
import Profile from './Todos'
import { setFilterCurrent, setFilterDone } from '../../redux/modules/todos'


type StateProps = {
  // Always put the asterisk or it assumes 'empty'
  todos: List<*>,
  filter: string,
}

function mapStateToProps (state: GlobalReducerState): StateProps {

  const { todos } = state

  return {
    todos: todos.get('list'),
    filter: todos.get('filter'),
  }

}


type ActionProps = {
  setFilterCurrent: Function,
  setFilterDone: Function,
}

export default connect(mapStateToProps, {
  setFilterCurrent,
  setFilterDone,
})(Profile)

export type ReduxProps = StateProps & ActionProps
