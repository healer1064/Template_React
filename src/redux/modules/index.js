// @flow

// Can replace rxjs import with individual operators with code-splitting in the
// future if bundle size becomes an issue.
// https://redux-observable.js.org/docs/Troubleshooting.html
import 'rxjs'
import { combineReducers } from 'redux'
import { combineEpics } from 'redux-observable'
import { routerReducer } from 'react-router-redux'
import env from './env'
import init from './init'
import profile, { fetchProfileEpic } from './profile'
import todos from './todos'


// redux-observable
// Full example: http://jsbin.com/jexomi/edit?js,output
export const rootEpic = combineEpics(
  fetchProfileEpic,
)

export const rootReducer = combineReducers({
  env,
  init,
  profile,
  todos,
  routing: routerReducer,
})
