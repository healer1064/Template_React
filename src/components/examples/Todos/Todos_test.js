// @flow
/* eslint-disable max-len */

// Enzyme docs:
// http://airbnb.io/enzyme/docs/api/index.html

import React from 'react'
import { shallow } from 'enzyme'
import { initialState } from 'src/redux/modules/todos'
import Todos from './Todos'

const mockProps = {
  setFilterCurrent: () => {},
  setFilterDone: () => {},
  getTodos: () => {},
}

it('<Todos> renders the Todos content', () => {

  const todos = initialState.get('list')
  const filter = initialState.get('filter')

  const wrapper = shallow(
    <Todos
      {...mockProps}
      todos={todos}
      filter={filter}
    />
  )

  expect(wrapper).toMatchSnapshot()

})
