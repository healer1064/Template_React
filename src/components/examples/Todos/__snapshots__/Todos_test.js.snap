// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todos> renders the Todos content 1`] = `
<div
  className="todos"
>
  <div
    className="modal"
  >
    Todos page
    <ul />
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/home"
    >
      Back
    </Link>
    <button
      onClick={[Function]}
    >
      Filter Done
    </button>
    <button
      onClick={[Function]}
    >
      Filter Current
    </button>
  </div>
</div>
`;
